### ISSUE TITLE
"no-unexpected-multiline": add support for chained method calls named by a symbol

### ISSUE THREAD

**Issue by Lcfvs** [eslint/eslint#12550]
**What rule do you want to change?**
no-unexpected-multiline

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior

**Please provide some example code that this change will affect:**

```js
const method = Symbol('methodDescription')

const obj = {
  [method] () {
    // do something
    return this
  }
}

obj
  [method]()
  [method]()
  [method]()
  [method]()
```

**What does the rule currently do for this code?**
Displays an error

**What will the rule do after it's changed?**
Treat chained symbol/eval'd method calls like the simple ones

**Are you willing to submit a pull request to implement this change?**
I don't know how without any breaking changes

**Comment by kaicataldo** [comment]
For ease of evaluation, here’s the reported example in our [demo](https://eslint.org/demo#eyJ0ZXh0Ijoib2JqXG4gIFttZXRob2RdKClcbiAgW21ldGhvZF0oKVxuICBbbWV0aG9kXSgpXG4gIFttZXRob2RdKCkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Mn0sImVudiI6e319fQ==).

**Comment by kaicataldo** [comment]
Thanks for reporting this. The reported case does seem like a case where the rule shouldn’t flag this (note that it’s about computed properties rather than related to Symbols). That being said, I’m not sure we should change the behavior for the following since it’s unclear the intention is:

```js
obj()
  ['a']()
```

**Comment by Lcfvs** [comment]
Hi @kaicataldo 

In your first example, the method isn't called on the object.

The second is fine, but may be problematic where we need to chain several method calls.

The real question is here: why to (not) support it, computed or not?

**Comment by kaicataldo** [comment]
I just clarified my example. The intention wasn’t to tie it to the original example - sorry for the confusion!

> The real question is here: why to (not) support it, computed or not?

The purpose of the rule is to disallow patterns that are ambiguous when not using semicolons. I think that since the intention could be either with the example above, that it’s correct to warn on it. 

From the docs (emphasis my own):
In the exceptions where a newline does not end a statement, a typing mistake to omit a semicolon causes two unrelated consecutive lines to be interpreted as one expression. Especially for a coding style without semicolons, readers might overlook the mistake. __Although syntactically correct, the code might throw exceptions when it is executed.__

**Comment by Lcfvs** [comment]
No worries. ;)

But what's the recommanded way to write it, please? Store each intermediary result isn't elegant and very verbose, especially when we are manipulating some immutables.

**Comment by kaicataldo** [comment]
I think the only option is disabling the rule for those lines.

Another ambiguous example:
```js
const foo = bar || baz
[method]()

// could be

const foo = bar || baz
[qux]()

// or

const foo = bar || baz[qux]();
```

Let's see what others have to say.

**Comment by Lcfvs** [comment]
I see the ambiguities.

But my future library is fully symbol-based, then enforce to disable a rule for a valid code by my library users is a few problematic.

It risks to make adverse publicity to my lib. :|

**Comment by Mouvedia** [comment]
related: #11197
